### PR TITLE
RFAI-04: Proposal revision data model and compiler contract

### PR DESCRIPTION
**Pull request by Chris0Jeky**
## Summary

Foundational data model slice for #976 (RFAI-04: Typed ProposalCompiler and revision-backed edit-before-approve flow).

- **ProposalRevision entity**: Immutable revision of an automation proposal. References the original proposal via FK. Contains revision number (monotonic, 1-based), editor identity, revised payload (full JSON snapshot -- never destructively overwrites original), timestamp, and reason. Private setters enforce immutability after construction.
- **ProposalOutcome entity**: Records content-free decision events with OutcomeType enum (Approved, EditedThenApproved, Rejected, Ignored).
- **IProposalCompiler interface**: Typed compiler contract in Domain layer with CompilerValidationResult (success/failure with risk aggregation).
- **OperationRisk value object**: Immutable risk level + reason with value equality semantics.
- **UnsupportedOperationFailure**: Structured failure for operations the compiler cannot process.
- **Application layer**: IProposalRevisionRepository (with latest-revision and next-number queries), ProposalRevisionDtos, IProposalRevisionService with create/list/get-latest.
- **Infrastructure**: EF Core migration creating ProposalRevisions and ProposalOutcomes tables with FK to AutomationProposals, unique index on (ProposalId, RevisionNumber). ProposalRevisionRepository, EF configurations, DI registration, DbContext and UnitOfWork updates.
- **70 new domain tests**: ProposalRevision creation/immutability/validation, revision chain integrity (latest resolution, ordering, no-revisions case, many-revisions), ProposalOutcome and OutcomeType coverage, CompilerValidationResult success/failure/aggregation, OperationRisk and UnsupportedOperationFailure value object semantics.

### Key design decisions

1. **Revisions never overwrite**: Each revision stores a full JSON snapshot. The original proposal payload on AutomationProposal is preserved. Latest effective payload is resolved via `GetLatestByProposalIdAsync`.
2. **Immutability**: ProposalRevision and ProposalOutcome use private setters throughout. No mutation methods exist after construction.
3. **Revision chain integrity**: Unique DB constraint on (ProposalId, RevisionNumber) prevents orphaned or duplicate revisions. RevisionNumber is auto-assigned via `GetNextRevisionNumberAsync`.
4. **OutcomeType is content-free**: Captures what happened (Approved, EditedThenApproved, Rejected, Ignored), not why.

### What awaits upstream

- Compiler wiring (concrete IProposalCompiler implementations) awaits upstream merges
- Frontend review card editing (edit-before-approve UI) awaits this data model landing
- API endpoints for revision CRUD will follow in a subsequent slice

## Test plan

- [x] `dotnet build backend/Taskdeck.sln -c Release` -- 0 errors
- [x] `dotnet test backend/Taskdeck.sln -c Release -m:1` -- 5,130 tests pass (0 failures)
- [x] Domain tests cover all new entities, value objects, and edge cases (70 new tests)
- [x] All existing fake UnitOfWork implementations updated for new interface member
- [x] EF migration generates correct tables, FKs, and indices

Closes: contributes to #976

### CODE REVIEW

**Comment by Chris0Jeky** [comment]
## Self-Review Findings

### 1. Immutability guarantees
**PASS.** All properties on ProposalRevision and ProposalOutcome use `private set`. No mutation methods exist after construction. The Entity base `Touch()` is protected but never called post-construction in these entities. Only EF Core's change tracker could theoretically mutate via reflection, which is inherent to the ORM pattern and acceptable.

### 2. Revision chain integrity
**PASS with note.** FK constraint (ProposalId -> AutomationProposals.Id) with Cascade delete prevents orphaned revisions. Unique index on (ProposalId, RevisionNumber) prevents duplicate revision numbers. `GetNextRevisionNumberAsync` uses `MAX(RevisionNumber) + 1` which is safe for single-writer scenarios. **Note:** Concurrent revision creation could race on the next revision number -- the unique index will reject duplicates, and the caller should retry. This is acceptable for the foundational data model slice; if high-concurrency editing is needed later, a pessimistic lock or optimistic retry can be added at the service layer.

### 3. Latest-revision resolution edge cases
**PASS.** Tested all three cases:
- No revisions: returns null, service returns `Result.Success(null)` -- original proposal payload is the effective payload.
- Single revision: returns that revision.
- Many revisions: `OrderByDescending(RevisionNumber).FirstOrDefault()` correctly returns highest.

### 4. EF migration correctness
**PASS.** ProposalRevisions and ProposalOutcomes tables created with correct column types, FKs to AutomationProposals with Cascade delete, unique index on (ProposalId, RevisionNumber), and separate indices on ProposalId, EditorUserId, and DecidedByUserId. Down migration drops both tables.

### 5. Outcome types content-free
**PASS.** OutcomeType enum has exactly 4 values (Approved, EditedThenApproved, Rejected, Ignored). ProposalOutcome stores only the type, deciding user, and timestamp -- no reason or rationale field. Content-free by design.

### Deferred items
- Concrete IProposalCompiler implementations (compiler wiring) await upstream merges.
- Frontend review card editing (edit-before-approve UI) awaits this data model.
- API endpoints for revision CRUD will be a subsequent slice.

**Comment by Chris0Jeky** [comment]
## Adversarial Review (Round 2)

### Gemini Findings -- All 8 Addressed

All 8 MEDIUM findings from Gemini have been fixed in commit d300c6f8:

| # | File | Line | Finding | Fix |
|---|------|------|---------|-----|
| 1 | `ProposalRevisionDtos.cs` | 9 | `RevisedAt` uses `DateTime` | Changed to `DateTimeOffset` |
| 2 | `ProposalRevisionDtos.cs` | 26 | `DecidedAt` uses `DateTime` | Changed to `DateTimeOffset` |
| 3 | `ProposalRevisionService.cs` | 44 | Race condition on `SaveChangesAsync` | Documented the race window with rationale (see below) |
| 4 | `ProposalRevisionService.cs` | 65 | Redundant `as IReadOnlyList<T>` cast | Removed; use explicit type parameter on `Result.Success<>` |
| 5 | `ProposalOutcome.cs` | 23 | `DecidedAt` uses `DateTime` | Changed to `DateTimeOffset` |
| 6 | `ProposalOutcome.cs` | 45 | Uses `DateTime.UtcNow` | Changed to `DateTimeOffset.UtcNow` |
| 7 | `ProposalRevision.cs` | 30 | `RevisedAt` uses `DateTime` | Changed to `DateTimeOffset` |
| 8 | `ProposalRevision.cs` | 65 | Uses `DateTime.UtcNow` | Changed to `DateTimeOffset.UtcNow` |

Migration, migration designer, model snapshot, and tests all updated to match.

### Race Condition Analysis (Finding #3)

The `GetNextRevisionNumberAsync` + `SaveChangesAsync` sequence has a TOCTOU race: two concurrent callers could read the same next revision number. The DB unique constraint on `(ProposalId, RevisionNumber)` prevents data corruption. However, the `DbUpdateException` from the unique constraint violation is **not** caught by the Application layer (doing so would require importing EF Core, violating GP-01 layer boundaries confirmed by Architecture.Tests). The UnitOfWork only handles Notification and UserPreference unique conflicts today.

**Decision**: Document the race window rather than introduce fragile exception-message sniffing. Rationale:
- Taskdeck is local-first; concurrent revision creation on the same proposal is extremely unlikely
- The DB constraint prevents data corruption (safety preserved)
- A proper fix belongs in UnitOfWork (Infrastructure layer) if this ever becomes a real issue
- Architecture tests enforce that Application must not reference `Microsoft.EntityFrameworkCore`

### Additional Adversarial Findings

**No blocking issues found.** The code is well-structured. Additional observations:

1. **RevisedPayload has no max length constraint** -- The EF configuration does not set a `HasMaxLength` on `RevisedPayload`. For SQLite TEXT columns this is fine (no practical limit), but if the project ever migrates to SQL Server or PostgreSQL, unbounded JSON columns could cause issues. Acceptable for now given the SQLite-first architecture.

2. **IProposalCompiler in Entities namespace** -- The interface lives in `Domain/Entities/` alongside entities. While this is valid (it's a domain contract), it might be clearer in a `Domain/Contracts/` or `Domain/Interfaces/` folder. Low priority, no functional impact.

3. **Value objects (OperationRisk, UnsupportedOperationFailure) are in Entities namespace** -- These are value objects, not entities. Same namespace concern as above. No functional impact.

4. **No max length on RevisedPayload in domain validation** -- The domain entity validates non-empty but does not cap the JSON payload size. If a caller sends a multi-MB payload, it will be persisted. Consider adding a domain-level max length if this is exposed via API.

5. **ProposalOutcome allows multiple outcomes per proposal** -- The `Outcomes` collection on `AutomationProposal` is a list, not a single value. This is by design for audit trails, but callers should be aware that creating a second outcome for an already-decided proposal is currently unconstrained at the domain level.

### Verification

- `dotnet build backend/Taskdeck.sln -c Release`: 0 errors, 22 warnings (all pre-existing)
- 70 domain tests + 1 API test: all passing
- 8 architecture tests: all passing (confirms no layer boundary violations)

File: backend/src/Taskdeck.Application/Services/ProposalRevisionService.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+            var revision = new ProposalRevision(
+                dto.ProposalId,
+                nextRevisionNumber,
+                dto.EditorUserId,
+                dto.RevisedPayload,
+                dto.Reason);
+
+            await _unitOfWork.ProposalRevisions.AddAsync(revision, cancellationToken);
+            await _unitOfWork.SaveChangesAsync(cancellationToken);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/ProposalRevisionService.cs:44]: Fixed in `414e5444`. `UnitOfWork` now maps the `ProposalRevisions(ProposalId, RevisionNumber)` unique-constraint collision to `DomainException(ErrorCodes.Conflict)`, so `ProposalRevisionService` returns a retryable conflict result without taking an EF Core dependency in Application. Added coverage at both boundaries:

- `ProposalRevisionServiceTests.CreateRevisionAsync_ReturnsConflict_WhenSaveChangesDetectsRevisionNumberRace`
- `ProposalRevisionRaceTests.SaveChangesAsync_DuplicateProposalRevisionNumber_ThrowsDomainConflict`

Verified with targeted Application/API tests, Architecture.Tests, and the ProposalRevision solution filter.

File: backend/src/Taskdeck.Application/DTOs/ProposalRevisionDtos.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+namespace Taskdeck.Application.DTOs;
+
+public record ProposalRevisionDto(
+    Guid Id,
+    Guid ProposalId,
+    int RevisionNumber,
+    Guid EditorUserId,
+    string RevisedPayload,
+    DateTime RevisedAt,
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/DTOs/ProposalRevisionDtos.cs]: Already fixed on the current branch by `d300c6f8`: `ProposalRevisionDto.RevisedAt` is `DateTimeOffset`. Re-verified this branch with the targeted `ProposalRevisionServiceTests` run, which passed.

File: backend/src/Taskdeck.Application/DTOs/ProposalRevisionDtos.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    string RevisedPayload,
+    string Reason
+);
+
+public record ProposalOutcomeDto(
+    Guid Id,
+    Guid ProposalId,
+    string OutcomeType,
+    Guid DecidedByUserId,
+    DateTime DecidedAt,
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/DTOs/ProposalRevisionDtos.cs]: Already fixed on the current branch by `d300c6f8`: `ProposalOutcomeDto.DecidedAt` is `DateTimeOffset`. Re-verified with targeted ProposalRevision tests; no code change needed in this pass.

File: backend/src/Taskdeck.Application/Services/ProposalRevisionService.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        CancellationToken cancellationToken = default)
+    {
+        var proposal = await _unitOfWork.AutomationProposals.GetByIdAsync(proposalId, cancellationToken);
+        if (proposal == null)
+            return Result.Failure<IReadOnlyList<ProposalRevisionDto>>(ErrorCodes.NotFound, "Proposal not found");
+
+        var revisions = await _unitOfWork.ProposalRevisions
+            .GetByProposalIdAsync(proposalId, cancellationToken);
+
+        var dtos = revisions.Select(MapToDto).ToList() as IReadOnlyList<ProposalRevisionDto>;
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/ProposalRevisionService.cs]: Already fixed on the current branch by `d300c6f8`: the redundant cast was removed and `List<ProposalRevisionDto>` is returned through the `IReadOnlyList<ProposalRevisionDto>` result. Targeted service tests passed.

File: backend/src/Taskdeck.Domain/Entities/ProposalOutcome.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    public Guid ProposalId { get; private set; }
+
+    /// <summary>The type of decision outcome.</summary>
+    public OutcomeType OutcomeType { get; private set; }
+
+    /// <summary>The user who made the decision.</summary>
+    public Guid DecidedByUserId { get; private set; }
+
+    /// <summary>When the decision was recorded (UTC).</summary>
+    public DateTime DecidedAt { get; private set; }
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/ProposalOutcome.cs]: Already fixed on the current branch by `d300c6f8`: `ProposalOutcome.DecidedAt` is `DateTimeOffset`. No additional code change was needed; targeted revision service/API race tests passed.

File: backend/src/Taskdeck.Domain/Entities/ProposalOutcome.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            throw new DomainException(ErrorCodes.ValidationError, "ProposalId cannot be empty");
+        if (decidedByUserId == Guid.Empty)
+            throw new DomainException(ErrorCodes.ValidationError, "DecidedByUserId cannot be empty");
+        if (!Enum.IsDefined(typeof(OutcomeType), outcomeType))
+            throw new DomainException(ErrorCodes.ValidationError, $"Invalid OutcomeType: {outcomeType}");
+
+        ProposalId = proposalId;
+        OutcomeType = outcomeType;
+        DecidedByUserId = decidedByUserId;
+        DecidedAt = DateTime.UtcNow;
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/ProposalOutcome.cs]: Already fixed on the current branch by `d300c6f8`: `ProposalOutcome` assigns `DecidedAt = DateTimeOffset.UtcNow`. Re-verified the branch with targeted ProposalRevision tests.

File: backend/src/Taskdeck.Domain/Entities/ProposalRevision.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    /// <summary>
+    /// The revised proposal payload as JSON. This is the full snapshot of the
+    /// edited operations, not a diff. The original proposal payload is preserved
+    /// on the AutomationProposal entity itself.
+    /// </summary>
+    public string RevisedPayload { get; private set; } = string.Empty;
+
+    /// <summary>Timestamp when the revision was created (UTC).</summary>
+    public DateTime RevisedAt { get; private set; }
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/ProposalRevision.cs]: Already fixed on the current branch by `d300c6f8`: `ProposalRevision.RevisedAt` is `DateTimeOffset`, consistent with the entity audit timestamps. Targeted tests passed.

File: backend/src/Taskdeck.Domain/Entities/ProposalRevision.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            throw new DomainException(ErrorCodes.ValidationError, "Reason cannot be empty");
+        if (reason.Length > 500)
+            throw new DomainException(ErrorCodes.ValidationError, "Reason cannot exceed 500 characters");
+
+        ProposalId = proposalId;
+        RevisionNumber = revisionNumber;
+        EditorUserId = editorUserId;
+        RevisedPayload = revisedPayload;
+        Reason = reason;
+        RevisedAt = DateTime.UtcNow;
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Domain/Entities/ProposalRevision.cs]: Already fixed on the current branch by `d300c6f8`: `ProposalRevision` assigns `RevisedAt = DateTimeOffset.UtcNow`. Re-verified with targeted service/API race tests.
